kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,382 @@
+// database was modified by a write transaction.
+func (tx *readWriteTx) Rollback() error {
+	// If the transaction has been closed roolback will fail.
+	if !tx.active {
+		return walletdb.ErrTxClosed
+	}
+
+	err := tx.tx.Rollback()
+
+	// Unlock the transaction regardless of the error result.
+	tx.active = false
+	tx.locker.Unlock()
+	return err
+}
+
+// ReadWriteBucket opens the root bucket for read/write access.  If the
+// bucket described by the key does not exist, nil is returned.
+func (tx *readWriteTx) ReadWriteBucket(key []byte) walletdb.ReadWriteBucket {
+	if len(key) == 0 {
+		return nil
+	}
+
+	table := tx.toTableName(key)
+
+	// If the key can't be mapped to a table name, open the bucket from the
+	// hex table.
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+		return bucket.NestedReadWriteBucket(key)
+	}
+
+	// Verify that the table containing the specified top-level bucket
+	// exists.
+	var value int
+	err := tx.QueryRow(
+		"SELECT 1 FROM pg_catalog.pg_tables "+
+			"WHERE schemaname='public' AND tablename=$1",
+		table,
+	).Scan(&value)
+
+	switch {
+	case err == sql.ErrNoRows:
+		return nil
+
+	case err != nil:
+		panic(err)
+	}
+
+	return newReadWriteBucket(tx, table, nil)
+}
+
+// getCreateTableSql returns the create statements to set up a kv table in
+// postgres. Every row points to the bucket that it is one via its parent_id
+// field. A NULL parent_id means that the key belongs to the upper-most bucket
+// in this table. A constraint on parent_id is enforcing referential integrity.
+//
+// Furthermore there is a <table>_p index on parent_id that is required for the
+// foreign key constraint.
+//
+// Finally there are unique indices on (parent_id, key) to prevent the same key
+// being present in a bucket more than once (<table>_up and <table>_unp). In
+// postgres, a single index wouldn't enforce the unique constraint on rows with
+// a NULL parent_id. Therefore two indices are defined.
+func getCreateTableSql(table string) string {
+	return `
+CREATE TABLE IF NOT EXISTS public.` + table + `
+(
+    key bytea NOT NULL,
+    value bytea,
+    parent_id bigint,
+    id bigserial PRIMARY KEY,
+    sequence bigint,
+    CONSTRAINT ` + table + `_parent FOREIGN KEY (parent_id)
+        REFERENCES public.` + table + ` (id)
+        ON UPDATE NO ACTION
+        ON DELETE CASCADE
+);
+
+CREATE INDEX IF NOT EXISTS ` + table + `_p
+    ON public.` + table + ` (parent_id);
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_up
+    ON public.` + table + `
+    (parent_id, key) WHERE parent_id IS NOT NULL;
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_unp 
+    ON public.` + table + ` (key) WHERE parent_id IS NULL;
+`
+}
+
+// CreateTopLevelBucket creates the top level bucket for a key if it
+// does not exist.  The newly-created bucket it returned.
+func (tx *readWriteTx) CreateTopLevelBucket(key []byte) (walletdb.ReadWriteBucket, error) {
```

Comment:
@guggero after letting this PR mature for a bit in my mind, I'd like to circle back to the 'split top level buckets into tables' question.

https://github.com/lightningnetwork/lnd/pull/5366#issuecomment-858552597

It looks cool, but I am not so sure anymore that this is the best option.

@yaslama already mentioned a few potential downsides: https://github.com/lightningnetwork/lnd/pull/5366#issuecomment-858294078

Originally performance was I think the main reason to split, but back then I already saw that it made no difference. At least on the benchmark.

You mentioned in your comment https://github.com/lightningnetwork/lnd/pull/5366#issuecomment-856937755:

> But having data spread out over multiple tables might also have other benefits

I don't think we zoomed in on those benefits anymore.

What I also noticed when playing around with the serializable locking model (better performance), is that it seems that serializable transactions that modify the actual db structure (create top level bucket) behave slightly different.

Taking everything into account, I think I'd favor reverting back to the single table model. This is a relatively small change because it mostly involves deleting code. Unless there are of course new (or forgotten...) reasons for sticking with the split tables.

Interested to hear your thoughts on this.

Reply:
>  is that it seems that serializable transactions that modify the actual db structure (create top level bucket) behave slightly different.

Can you elaborate on this?

Reply:
More background in the first answer of https://stackoverflow.com/questions/54351783/duplicate-key-value-violates-unique-constraint-postgres-error-when-trying-to-c

When updating the database structure itself concurrently, you can get a different type of error. I think it is fine if we just catch that error too and retry the transaction in case it happens - like we do for other serializable errors. But why even go there? Concurrent updates of the table structure are quite unusual and I'd rather stay away from if there is no real benefit.

Reply:
Added temporary commit to show delta when moving back to the single table model. Basically all logic was already there, because the `hex` table is a root bucket for the non-human-readable bucket key.

Reply:
I think my main concerns or arguments for splitting were:
1. More visibility (easily see what top level buckets there are by looking at the table names and make it easy to identify the large data sections of the DB by looking at the largest tables). But I guess that can also be done with a simple query instead (though that would need to be a recursive one to count all keys within a top level bucket).
2. More fine-grained tuning. I'm not very familiar with expression based indices and whether they can reach the same level of performance as per-table indices. I just assume that having multiple tables gives a database administrator more options in terms of identifying and addressing bottlenecks that relate to only a part of the data (for example the payments data or channel state). Being able to separate DDL from DML might be a valid concern for DBAs here though. And I can see how executing DDL dynamically can change the transactional behavior of the code.

I don't have very strong preference either way, though with the argument that using DDL within a transaction might lead to slightly different behavior and the fact that the code also gets simplified, I'd perhaps lean more toward rolling back to a single table.

Reply:
Rolled back to single table.